packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesModule.kt

```diff
@@ -68,143 +68,129 @@ class UpdatesModule : Module(), IUpdatesEventManagerObserver {
-                promise.resolve(
-                  Bundle().apply {
-                    putBoolean("isRollBackToEmbedded", true)
-                    putBoolean("isAvailable", false)
-                  }
-                )
-              }
+            is IUpdatesController.CheckForUpdateResult.NoUpdateAvailable -> {
+              promise.resolve(
+                Bundle().apply {
+                  putBoolean("isRollBackToEmbedded", false)
+                  putBoolean("isAvailable", false)
+                  putString("reason", result.reason.value)
+                }
+              )
+            }
 
-              is IUpdatesController.CheckForUpdateResult.UpdateAvailable -> {
-                promise.resolve(
-                  Bundle().apply {
-                    putBoolean("isRollBackToEmbedded", false)
-                    putBoolean("isAvailable", true)
-                    putString(
-                      "manifestString",
-                      result.update.manifest.toString()
-                    )
-                  }
-                )
-              }
+            is IUpdatesController.CheckForUpdateResult.RollBackToEmbedded -> {
+              promise.resolve(
+                Bundle().apply {
+                  putBoolean("isRollBackToEmbedded", true)
+                  putBoolean("isAvailable", false)
+                }
+              )
             }
-          }
 
-          override fun onFailure(exception: CodedException) {
-            promise.reject(exception)
+            is IUpdatesController.CheckForUpdateResult.UpdateAvailable -> {
+              promise.resolve(
+                Bundle().apply {
+                  putBoolean("isRollBackToEmbedded", false)
+                  putBoolean("isAvailable", true)
+                  putString(
+                    "manifestString",
+                    result.update.manifest.toString()
+                  )
+                }
+              )
+            }
           }
+        }.onFailure {
+          promise.reject(it as CodedException)
         }
-      )
+      }
     }
 
     AsyncFunction("fetchUpdateAsync") { promise: Promise ->
-      UpdatesController.instance.fetchUpdate(
-        object : IUpdatesController.ModuleCallback<IUpdatesController.FetchUpdateResult> {
-          override fun onSuccess(result: IUpdatesController.FetchUpdateResult) {
-            when (result) {
-              is IUpdatesController.FetchUpdateResult.ErrorResult -> {
-                promise.reject("ERR_UPDATES_FETCH", "Failed to download new update", result.error)
-              }
-
-              is IUpdatesController.FetchUpdateResult.Failure -> {
-                promise.resolve(
-                  Bundle().apply {
-                    putBoolean("isRollBackToEmbedded", false)
-                    putBoolean("isNew", false)
-                  }
-                )
-              }
+      moduleScope.launch {
```

Comment:
note that the `moduleScope` is now binded to Dispatchers.IO which is a thread pool under the hood. not sure if we like to have this behavior change. the other way is to use the coroutine support from AsyncFunction: https://docs.expo.dev/modules/module-api/#kotlin-coroutines-

that would bind the execution on the original expo-modules-core default queue

Reply:
it's not specific to `fetchUpdateAsync`, but also to the other functions in this pr. for example, what would happen if `reload` be called in parallel. that would probably depend on how we lock critical section. if we are confident for the UpdatesController to support parallel calls, then it would be fine to use thread pool here.

Reply:
Good point! As an aside, if we use the serial queue, doesn't that mean cancellation will only happen when the appContext is destroyed? That would only happen on reload

Reply:
correct. i’m not so concerned for cancellation because both AppContext moduleCoroutine and this UpdatesModule.moduleCoroutine are destroyed when reload. and react native reload is happened as a last step for Updates.reloadAsync. i’m more concerned for the steps before react native reload.

well, i think that’s fine either to use thread pool because you have some locks in UpdatesController. i’m leaving the final call for you